htpy/html2htpy.py

```diff
@@ -0,0 +1,284 @@
+
+    def serialize_python(self, shorthand_id_class: bool = False, format: bool = False):
+        o = ""
+
+        if len(self._collected) == 1:
+            o += _serialize(self._collected[0], shorthand_id_class)
+
+        else:
+            o += "["
+            for t in self._collected:
+                o += _serialize(t, shorthand_id_class) + ","
+            o = o[:-1] + "]"
+
+        if format:
+            try:
+                import black
+            except:
+                raise Exception(
+                    "Cannot import formatter. Please ensure black is installed."
+                )
+
+            return black.format_str(
+                o, mode=black.FileMode(line_length=80, magic_trailing_comma=False)
+            )
+        else:
+            return o
+
+
+def html2htpy(html: str, shorthand_id_class: bool = False, format: bool = False):
+    parser = HTPYParser()
+    parser.feed(html)
+
+    return parser.serialize_python(shorthand_id_class, format)
+
+
+def _convert_data_to_string(data: str):
+    _data = str(data)
+
+    # escape unescaped dblquote: " -> \"
+    _data = re.compile(r'(?<![\\])"').sub('\\"', _data)
+
+    template_string_pattern = re.compile(r"\{\{\s*(\w+)\s*\}\}")
+
+    has_jinja_pattern = re.search(template_string_pattern, _data)
+    if has_jinja_pattern:
+        # regex replaces these 3 cases:
+        # {{ var }} -> { var }
+        # { -> {{
+        # } -> }}
+        template_string_replace_pattern = re.compile(
+            r"(\{\{\s*(\w+)\s*\}\}|(?<![\{]){(?![\{])|(?<![\}])}(?![\}]))"
+        )
+
+        def replacer(match: re.Match[str]):
+            captured = match.group(1)
+
+            if captured.startswith("{{"):
+                return captured[1:-1]
+
+            if captured == "{":
+                return "{{"
+
+            return "}}"
+
+        _data = template_string_replace_pattern.sub(replacer, _data)
+        _data = 'f"' + _data + '"'
+    else:
+        _data = '"' + _data + '"'
+
+    return _data
+
+
+def _serialize(el: Tag | str, shorthand_id_class: bool):
+    if isinstance(el, Tag):
+        return el.serialize(shorthand_id_class=shorthand_id_class)
+    else:
+        return str(el)
+
+
+@dataclass
+class ConvertArgs:
+    shorthand: bool
+    format: bool
+
+
+def main():
+    parser = argparse.ArgumentParser(prog="html2htpy")
+
+    parser.add_argument(
+        "-s",
+        "--shorthand",
+        help="Use shorthand syntax for class and id attributes",
+        action="store_true",
+    )
+    parser.add_argument(
+        "-f",
+        "--format",
+        help="Format output code (requires black installed)",
+        action="store_true",
+    )
+    parser.add_argument(
+        "input",
+        type=argparse.FileType("r"),
+        nargs="?",
+        default=sys.stdin,
+        help="input html from file or stdin",
+    )
+
+    args = parser.parse_args()
+
+    if args.input == sys.stdin and select.select([sys.stdin], [], [], 0.1)[0]:
```

Comment:
what does the select call do here? would it work without it?

Reply:
I must admit I'm fairly light on this topic. The internet is my guide.

My understanding, however, is that it checks if there is input on the stdin stream (or if input shows up within a deadline of 0.1 seconds). On my system, without it, the command hangs if you run it without piping input (if you did not pass a file). With it, python detects if there is no input piped AND no file provided, so it can enter the else block, error out and exit.

see this SO-answer: [https://stackoverflow.com/a/3763257](https://stackoverflow.com/a/3763257)

According to SO answer this might only work on unix systems, and I've only tested it on linux.

If you have any advice I am open ears. Otherwise I can investigate this topic further tomorrow.

Reply:
Maybe the command hanging without input is actually correct behaviour. I see that grep behaves the same way, as also referenced in the SO post.

Reply:
Hanging is how most commands work and what is generally expected. It also allows for pasting things directly into a terminal (I use this trick often). I agree it may seem confusing and weird that it just hangs but that is the convention.

For instance, running `black -` and then typing `a=1<enter><ctrl+d>` looks like this:
```
$ black -
a=1
a = 1
reformatted -

All done! ✨ 🍰 ✨
1 file reformatted.
```

Reply:
Excellent, learning new things here! I removed the select call and it works like you'd expect!